### PR TITLE
[rl] Add batched RL training materials

### PR DESCRIPTION
**Pull request by joecummings**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #2837
* #2836
* #2835
* #2834
* __->__ #2851

### CODE REVIEW

File: torchtitan/experiments/rl/types.py [hunk 1]
```diff
@@ -43,3 +45,15 @@ class Episode:
     reward: float = 0.0
     group_id: str = ""
     advantage: float = 0.0
+
+
+@dataclass
+class TrainBatch:
```
Comment (joecummings) [inline torchtitan/experiments/rl/types.py:51]: Adding some structure for training. Probably not the final form.

File: torchtitan/experiments/rl/actors/trainer.py [hunk 2]
```diff
@@ -277,49 +278,45 @@ async def step(self, episodes: list[Episode]) -> dict:
             f"{os.getpid()=} PolicyTrainer starting step {self.policy_version} "
         )
 
-        advantages = torch.tensor([ep.advantage for ep in episodes])
+        local_batch = batches[self.dp_rank]
```
Comment (wwwjn) [inline torchtitan/experiments/rl/actors/trainer.py:281]: Can we only send the local batch to trainer?

File: torchtitan/experiments/rl/actors/trainer.py [hunk 3]
```diff
@@ -277,49 +278,45 @@ async def step(self, episodes: list[Episode]) -> dict:
...
-            self.model,
-            my_token_ids,
-            my_prompt_token_ids,
-            my_advantages,
-            ref_token_log_probs,
-            kl_coef=0.1,
+            ref_logits = self.ref_model(token_ids)
+            ref_logprobs = extract_logprobs_batched(ref_logits, token_ids)
+
+        response_mask = build_response_mask(
```
Comment (wwwjn) [inline torchtitan/experiments/rl/actors/trainer.py:296]: nit: we can add inline comments to indicate the tensor's shape

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py [hunk 4]
```diff
@@ -217,6 +312,9 @@ async def setup(
         self.generator_world_size = self._compute_world_size(
             config.generator.parallelism
         )
+        tp = config.trainer.parallelism
```
Comment (wwwjn) [inline torchtitan/experiments/rl/simple_grpo_sum_digits.py:315]:
```suggestion
        self.trainer_dp_degree = config.trainer.parallelism.get_mesh("batch")
```

https://github.com/pytorch/torchtitan/blob/refs/heads/main/torchtitan/trainer.py#L218

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py [hunk 5]
```diff
@@ -303,6 +401,7 @@ async def setup(
             "trainer",
             PolicyTrainer,
             config.trainer,
+            loss_fn=GRPOLoss(),
```
Comment (wwwjn) [inline torchtitan/experiments/rl/simple_grpo_sum_digits.py:404]: Can we put this `Loss` part of RLTrainer / PolicyTrainer's configuration?

File: torchtitan/experiments/rl/types.py (hunk 1 above)
Comment (wwwjn) [reply torchtitan/experiments/rl/types.py:51]: Why this is pre-collated? An alternative is to define `TrainBatch` shard and collated batch and ready to be trained on
